state/compatibility_test.go

```diff
@@ -98,81 +140,157 @@ func TestSaveLegacyAndLoadFinalizeBlock(t *testing.T) {
 	multiStore := NewMultiStore(stateDB, options, store)
 
 	// try with a complete ABCI Response
-	abciResponses := newV1Beta2ABCIResponses()
-	require.Equal(t, 1, len(abciResponses.DeliverTxs))
-	require.Equal(t, 1, len(abciResponses.BeginBlock.Events))
-	require.Equal(t, 1, len(abciResponses.EndBlock.Events))
-	err := multiStore.SaveABCIResponses(height, &abciResponses)
+	v1beta2ABCIResponses := newV1Beta2ABCIResponses()
+	err := multiStore.SaveABCIResponses(height, &v1beta2ABCIResponses)
 	require.NoError(t, err)
-	loadedABCIResponses, err := multiStore.LoadFinalizeBlockResponse(height)
+	finalizeBlockResponse, err := multiStore.LoadFinalizeBlockResponse(height)
 	require.NoError(t, err)
-	require.Equal(t, 1, len(loadedABCIResponses.TxResults))
-	require.Equal(t, abciResponses.DeliverTxs[0].String(), loadedABCIResponses.TxResults[0].String())
-	require.Equal(t, 2, len(loadedABCIResponses.Events))
-	require.Equal(t, len(abciResponses.BeginBlock.Events)+len(abciResponses.EndBlock.Events), len(loadedABCIResponses.Events))
+
+	// Test for not nil
+	require.NotNil(t, finalizeBlockResponse.TxResults)
+	require.NotNil(t, finalizeBlockResponse.Events)
+	require.NotNil(t, finalizeBlockResponse.ValidatorUpdates)
+	require.NotNil(t, finalizeBlockResponse.ConsensusParamUpdates)
+	require.Nil(t, finalizeBlockResponse.AppHash)
+
+	// Test for equality
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs), len(finalizeBlockResponse.TxResults))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Code, finalizeBlockResponse.TxResults[0].Code)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Data, finalizeBlockResponse.TxResults[0].Data)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Log, finalizeBlockResponse.TxResults[0].Log)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasWanted, finalizeBlockResponse.TxResults[0].GasWanted)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].GasUsed, finalizeBlockResponse.TxResults[0].GasUsed)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events), len(finalizeBlockResponse.TxResults[0].Events))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Type, finalizeBlockResponse.TxResults[0].Events[0].Type)
+	require.Equal(t, len(v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes), len(finalizeBlockResponse.TxResults[0].Events[0].Attributes))
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Key, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Key)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Events[0].Attributes[0].Value, finalizeBlockResponse.TxResults[0].Events[0].Attributes[0].Value)
+	require.Equal(t, v1beta2ABCIResponses.DeliverTxs[0].Codespace, finalizeBlockResponse.TxResults[0].Codespace)
```

Comment:
Excellent! Just one question (not sure about this): if you call
```
require.Equal(v1beta2ABCIResponses.DeliverTxs, finalizeBlockResponse.TxResults)
```
Won't that check all the values and all the fields in the values?

Reply:
Actually, that's what you're doing in line 211 below (with a different value for `v1beta2ABCIResponses`)

Reply:
I can't check them for equality straight away because they are based on different protos so the check fails (I tried that first), because even though structurally they might look the same their fields might come from a different version and the `Equals` fails

Reply:
```
=== RUN   TestSaveLegacyAndLoadFinalizeBlock
    compatibility_test.go:158: 
        	Error Trace:	/Users/andynogueira/Development/Go/src/github.com/cometbft/cometbft/state/compatibility_test.go:158
        	Error:      	Not equal: 
        	            	expected: []*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x140002e0380)})
        	            	actual  : []*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x140002e0400)})
        	Test:       	TestSaveLegacyAndLoadFinalizeBlock
--- FAIL: TestSaveLegacyAndLoadFinalizeBlock (0.00s)

Expected :[]*v1beta2.ResponseDeliverTx([]*v1beta2.ResponseDeliverTx{(*v1beta2.ResponseDeliverTx)(0x140002e0380)})
Actual   :[]*v1.ExecTxResult([]*v1.ExecTxResult{(*v1.ExecTxResult)(0x140002e0400)})
```